File: crates/markuplint-rules/src/aria/accname/element_names.rs
//! `AccName` Step 2E: Element-specific name computation (HTML-AAM §4.1).

#![allow(clippy::too_many_arguments)]

use std::collections::HashSet;

use markuplint_dom::arena::{DomArena, NodeId};
use markuplint_dom::helpers as dom;

use super::helpers::{
    collect_text_content, find_child_by_local_name, get_input_type, is_svg_element, resolve_name_from_content,
};
use super::label_steps::resolve_label_text;
use super::{AccnameResolver, AccnameResult, AccnameSource};

const TEXT_INPUT_TYPES: &[&str] = &["text", "search", "tel", "url", "email", "password", "number"];

/// Step 2E: Get element-specific accessible name.
pub fn get_element_specific_name(
    arena: &DomArena,
    node_id: NodeId,
    resolver: &dyn AccnameResolver,
    visited: &HashSet<String>,
    in_labelledby_traversal: bool,
    computing: &mut HashSet<NodeId>,
) -> Option<AccnameResult> {
    let el = arena.get(node_id)?.as_element()?;
    let tag = el.base.node_name.as_str();

    if is_svg_element(arena, node_id) {
        return handle_svg_element(arena, node_id);
    }

    match tag {
        "input" => handle_input(arena, node_id, resolver, visited, in_labelledby_traversal, computing),
        "textarea" | "select" | "meter" | "progress" | "output" => {
            handle_labelable_with_title(arena, node_id, resolver, visited, in_labelledby_traversal, computing)
        }
        "button" => handle_button(arena, node_id, resolver, visited, in_labelledby_traversal, computing),
        "fieldset" => handle_fieldset(arena, node_id, resolver, visited, in_labelledby_traversal, computing),
        "table" => handle_table(arena, node_id, resolver, visited, in_labelledby_traversal, computing),
        "img" => handle_img(arena, node_id),
        "area" => handle_area(arena, node_id),
        "figure" | "iframe" => handle_title_only(arena, node_id),
        "summary" => handle_summary(arena, node_id, resolver, visited, in_labelledby_traversal, computing),
        "a" => {
            if dom::has_attr(arena, node_id, "href") {
                handle_anchor(arena, node_id, resolver, visited, in_labelledby_traversal, computing)
            } else {
                None
            }
        }
        _ => None,
    }
}

fn handle_input(
    arena: &DomArena,
    node_id: NodeId,
    resolver: &dyn AccnameResolver,
    visited: &HashSet<String>,
    in_labelledby: bool,
    computing: &mut HashSet<NodeId>,
) -> Option<AccnameResult> {
    let input_type = get_input_type(arena, node_id);

    if input_type == "hidden" {
        return Some(AccnameResult::empty());
    }

    if TEXT_INPUT_TYPES.contains(&input_type.as_str()) {
        return handle_labelable_with_title(arena, node_id, resolver, visited, in_labelledby, computing);
    }

    match input_type.as_str() {
        "button" | "submit" | "reset" => {
            handle_input_button(arena, node_id, &input_type, resolver, visited, in_labelledby, computing)
        }
        "image" => handle_input_image(arena, node_id, resolver, visited, in_labelledby, computing),
        _ => {
            if let Some(r) = resolve_label_text(arena, node_id, resolver, visited, in_labelledby, computing) {
                return Some(r);
            }
            title_attr(arena, node_id)
        }
    }
}

fn handle_input_button(
    arena: &DomArena,
    node_id: NodeId,
    input_type: &str,
    resolver: &dyn AccnameResolver,
    visited: &HashSet<String>,
    in_labelledby: bool,
    computing: &mut HashSet<NodeId>,
) -> Option<AccnameResult> {
    if let Some(r) = resolve_label_text(arena, node_id, resolver, visited, in_labelledby, computing) {
        return Some(r);
    }
    if let Some(v) = dom::get_attr_value(arena, node_id, "value")
        && !v.is_empty()
    {
        return Some(AccnameResult::new(v, AccnameSource::Value));
    }
    if let Some(r) = title_attr(arena, node_id) {
        return Some(r);
    }
    let default = match input_type {
        "submit" => "Submit",
        "reset" => "Reset",
        _ => "",
    };
    if default.is_empty() {
        None
    } else {
        Some(AccnameResult::new(default, AccnameSource::Default))
    }
}

#[allow(clippy::unnecessary_wraps)]
fn handle_input_image(
    arena: &DomArena,
    node_id: NodeId,
    resolver: &dyn AccnameResolver,
    visited: &HashSet<String>,
    in_labelledby: bool,
    computing: &mut HashSet<NodeId>,
) -> Option<AccnameResult> {
    if let Some(r) = resolve_label_text(arena, node_id, resolver, visited, in_labelledby, computing) {
        return Some(r);
    }
    if let Some(alt) = dom::get_attr_value(arena, node_id, "alt")
        && !alt.is_empty()
    {
        return Some(AccnameResult::new(alt, AccnameSource::Alt));
    }
    if let Some(r) = title_attr(arena, node_id) {
        return Some(r);
    }
    Some(AccnameResult::new("Submit Query", AccnameSource::Default))
}

fn handle_labelable_with_title(
    arena: &DomArena,
    node_id: NodeId,
    resolver: &dyn AccnameResolver,
    visited: &HashSet<String>,
    in_labelledby: bool,
    computing: &mut HashSet<NodeId>,
) -> Option<AccnameResult> {
    if let Some(r) = resolve_label_text(arena, node_id, resolver, visited, in_labelledby, computing) {
        return Some(r);
    }
    if let Some(r) = title_attr(arena, node_id) {
        return Some(r);
    }
    placeholder_attr(arena, node_id)
}

fn handle_button(
    arena: &DomArena,
    node_id: NodeId,
    resolver: &dyn AccnameResolver,
    visited: &HashSet<String>,
    in_labelledby: bool,
    computing: &mut HashSet<NodeId>,
) -> Option<AccnameResult> {
    if let Some(r) = resolve_label_text(arena, node_id, resolver, visited, in_labelledby, computing) {
        return Some(r);
    }
    let content = resolve_name_from_content(arena, node_id, resolver, visited, in_labelledby, computing);
    if !content.is_empty() {
        return Some(AccnameResult::new(content, AccnameSource::Content));
    }
    title_attr(arena, node_id)
}

fn handle_fieldset(
    arena: &DomArena,
    node_id: NodeId,
    resolver: &dyn AccnameResolver,
    visited: &HashSet<String>,
    in_labelledby: bool,
    computing: &mut HashSet<NodeId>,
) -> Option<AccnameResult> {
    if let Some(legend_id) = find_child_by_local_name(arena, node_id, "legend") {
        let content = resolve_name_from_content(arena, legend_id, resolver, visited, in_labelledby, computing);
        if !content.is_empty() {
            return Some(AccnameResult::new(content, AccnameSource::Legend));
        }
    }
    title_attr(arena, node_id)
}

fn handle_table(
    arena: &DomArena,
    node_id: NodeId,
    resolver: &dyn AccnameResolver,
    visited: &HashSet<String>,
    in_labelledby: bool,
    computing: &mut HashSet<NodeId>,
) -> Option<AccnameResult> {
    if let Some(caption_id) = find_child_by_local_name(arena, node_id, "caption") {
        let content = resolve_name_from_content(arena, caption_id, resolver, visited, in_labelledby, computing);
        if !content.is_empty() {
            return Some(AccnameResult::new(content, AccnameSource::Caption));
        }
    }
    title_attr(arena, node_id)
}

fn handle_img(arena: &DomArena, node_id: NodeId) -> Option<AccnameResult> {
    if dom::has_attr(arena, node_id, "alt") {
        let alt = dom::get_attr_value(arena, node_id, "alt").unwrap_or("");
        return Some(AccnameResult {
            name: super::helpers::flatten_text(alt),
            source: Some(AccnameSource::Alt),
        });
    }
    title_attr(arena, node_id)
}

fn handle_area(arena: &DomArena, node_id: NodeId) -> Option<AccnameResult> {
    if let Some(alt) = dom::get_attr_value(arena, node_id, "alt")
        && !alt.is_empty()
    {
        return Some(AccnameResult::new(alt, AccnameSource::Alt));
    }
    title_attr(arena, node_id)
}

fn handle_summary(
    arena: &DomArena,
    node_id: NodeId,
    resolver: &dyn AccnameResolver,
    visited: &HashSet<String>,
    in_labelledby: bool,
    computing: &mut HashSet<NodeId>,
) -> Option<AccnameResult> {
    let content = resolve_name_from_content(arena, node_id, resolver, visited, in_labelledby, computing);
    if !content.is_empty() {
        return Some(AccnameResult::new(content, AccnameSource::Content));
    }
    title_attr(arena, node_id)
}

fn handle_anchor(
    arena: &DomArena,
    node_id: NodeId,
    resolver: &dyn AccnameResolver,
    visited: &HashSet<String>,
    in_labelledby: bool,
    computing: &mut HashSet<NodeId>,
) -> Option<AccnameResult> {
    let content = resolve_name_from_content(arena, node_id, resolver, visited, in_labelledby, computing);
    if !content.is_empty() {
        return Some(AccnameResult::new(content, AccnameSource::Content));
    }
    title_attr(arena, node_id)
}

fn handle_svg_element(arena: &DomArena, node_id: NodeId) -> Option<AccnameResult> {
    if let Some(title_id) = find_child_by_local_name(arena, node_id, "title") {
        let text = collect_text_content(arena, title_id);
        if !text.is_empty() {
            return Some(AccnameResult::new(text, AccnameSource::SvgTitle));
        }
    }
    None
}

fn handle_title_only(arena: &DomArena, node_id: NodeId) -> Option<AccnameResult> {
    title_attr(arena, node_id)
}

fn title_attr(arena: &DomArena, node_id: NodeId) -> Option<AccnameResult> {
    let title = dom::get_attr_value(arena, node_id, "title")?;
    if title.trim().is_empty() {
        return None;
    }
    Some(AccnameResult::new(title, AccnameSource::Title))
}

fn placeholder_attr(arena: &DomArena, node_id: NodeId) -> Option<AccnameResult> {
    let placeholder = dom::get_attr_value(arena, node_id, "placeholder")?;
    if placeholder.trim().is_empty() {
        return None;
    }
    Some(AccnameResult::new(placeholder, AccnameSource::Placeholder))
}

File: packages/@markuplint/ml-spec/src/algorithm/aria/accname/element-names.ts
/* eslint-disable @typescript-eslint/prefer-readonly-parameter-types -- AccnameElement wraps mutable DOM types */

import type { AccnameElement, AccnameResolver, AccnameResult } from './types.js';

import {
	DEFAULT_IMAGE_LABEL,
	DEFAULT_RESET_LABEL,
	DEFAULT_SUBMIT_LABEL,
	TEXT_INPUT_TYPES,
} from '../../../const/index.js';
import { findChildByLocalName, getInputType, isSvgElement, makeResult, resolveNameFromContent } from './helpers.js';
import { resolveLabelText } from './label-steps.js';

/**
 * Implements AccName 1.2 §4.3.2 Step 2E: for elements that have a native
 * host language text alternative, use that alternative. The specific rules
 * for each HTML element are defined in HTML-AAM §4.1.
 *
 * Returns null if no element-specific rule applies, letting the caller
 * fall through to name-from-content (Step 2F) or title fallback (Step 2I).
 *
 * @see https://www.w3.org/TR/accname-1.2/#computation-steps — AccName 1.2 §4.3.2 Step 2E
 * @see https://www.w3.org/TR/html-aam-1.0/#accessible-name-and-description-computation — HTML-AAM §4.1
 */
export function getElementSpecificName(
	el: AccnameElement,
	resolver: AccnameResolver,
	visited: ReadonlySet<string>,
	computeFn: (
		el: AccnameElement,
		resolver: AccnameResolver,
		inLabelledbyTraversal: boolean,
		visited: ReadonlySet<string>,
	) => AccnameResult,
	inLabelledbyTraversal: boolean,
): AccnameResult | null {
	if (isSvgElement(el)) {
		return handleSvgElement(el);
	}

	const { localName } = el;

	switch (localName) {
		case 'input': {
			return handleInput(el, resolver, visited, computeFn, inLabelledbyTraversal);
		}
		case 'textarea':
		case 'select':
		case 'meter':
		case 'progress':
		case 'output': {
			return handleLabelableWithTitle(el, resolver, visited, computeFn, inLabelledbyTraversal);
		}
		case 'button': {
			return handleButton(el, resolver, visited, computeFn, inLabelledbyTraversal);
		}
		case 'fieldset': {
			return handleFieldset(el, resolver, visited, computeFn, inLabelledbyTraversal);
		}
		case 'table': {
			return handleTable(el, resolver, visited, computeFn, inLabelledbyTraversal);
		}
		case 'img': {
			return handleImg(el);
		}
		case 'area': {
			return handleArea(el);
		}
		case 'figure': {
			return handleFigure(el);
		}
		case 'summary': {
			return handleSummary(el, resolver, visited, computeFn, inLabelledbyTraversal);
		}
		case 'a': {
			if (el.hasAttribute('href')) {
				return handleAnchor(el, resolver, visited, computeFn, inLabelledbyTraversal);
			}
			return null;
		}
		case 'iframe': {
			return handleTitleOnly(el);
		}
		default: {
			// All other elements (tr, td, th, section, div, span, p, h1-h6, etc.)
			// only get name from title (handled by caller's title fallback)
			return null;
		}
	}
}

/**
 * HTML-AAM §4.1 (input[text-like], textarea, select, meter, progress, output).
 * Name sources: label → title → placeholder.
 * @see https://www.w3.org/TR/html-aam-1.0/#el-input-text
 */
function handleLabelableWithTitle(
	el: AccnameElement,
	resolver: AccnameResolver,
	visited: ReadonlySet<string>,
	computeFn: (
		el: AccnameElement,
		resolver: AccnameResolver,
		inLabelledbyTraversal: boolean,
		visited: ReadonlySet<string>,
	) => AccnameResult,
	inLabelledbyTraversal: boolean,
): AccnameResult | null {
	const labelResult = resolveLabelText(el, resolver, visited, computeFn, inLabelledbyTraversal);
	if (labelResult) {
		return labelResult;
	}

	const title = el.getAttribute('title');
	if (title?.trim()) {
		return makeResult(title, 'title');
	}

	const placeholder = el.getAttribute('placeholder');
	if (placeholder?.trim()) {
		return makeResult(placeholder, 'placeholder');
	}

	return null;
}

/**
 * HTML-AAM §4.1 (input — dispatches by type).
 * Text-like types: label → title → placeholder.
 * Button/submit/reset: label → value → default label.
 * Image: label → alt → title → default label.
 * Hidden: always empty.
 * Checkbox/radio: label → title.
 * @see https://www.w3.org/TR/html-aam-1.0/#el-input-text
 */
function handleInput(
	el: AccnameElement,
	resolver: AccnameResolver,
	visited: ReadonlySet<string>,
	computeFn: (
		el: AccnameElement,
		resolver: AccnameResolver,
		inLabelledbyTraversal: boolean,
		visited: ReadonlySet<string>,
	) => AccnameResult,
	inLabelledbyTraversal: boolean,
): AccnameResult | null {
	const type = getInputType(el);

	if (TEXT_INPUT_TYPES.has(type)) {
		return handleLabelableWithTitle(el, resolver, visited, computeFn, inLabelledbyTraversal);
	}

	if (type === 'button' || type === 'submit' || type === 'reset') {
		return handleInputButton(el, type, resolver, visited, computeFn, inLabelledbyTraversal);
	}

	if (type === 'image') {
		return handleInputImage(el, resolver, visited, computeFn, inLabelledbyTraversal);
	}

	if (type === 'hidden') {
		return makeResult('', null);
	}

	// checkbox, radio: label -> title
	return handleLabelableWithTitle(el, resolver, visited, computeFn, inLabelledbyTraversal);
}

/**
 * HTML-AAM §4.1 (input[type=button/submit/reset]).
 * Name sources: label → value → title → default label (submit="Submit", reset="Reset").
 * @see https://www.w3.org/TR/html-aam-1.0/#el-input-button
 */
function handleInputButton(
	el: AccnameElement,
	type: string,
	resolver: AccnameResolver,
	visited: ReadonlySet<string>,
	computeFn: (
		el: AccnameElement,
		resolver: AccnameResolver,
		inLabelledbyTraversal: boolean,
		visited: ReadonlySet<string>,
	) => AccnameResult,
	inLabelledbyTraversal: boolean,
): AccnameResult | null {
	const labelResult = resolveLabelText(el, resolver, visited, computeFn, inLabelledbyTraversal);
	if (labelResult) {
		return labelResult;
	}

	const value = el.getAttribute('value');
	if (value?.trim()) {
		return makeResult(value, 'value');
	}

	const title = el.getAttribute('title');
	if (title?.trim()) {
		return makeResult(title, 'title');
	}

	// Default label for submit/reset buttons
	if (type === 'submit') {
		return makeResult(DEFAULT_SUBMIT_LABEL, 'default');
	}
	if (type === 'reset') {
		return makeResult(DEFAULT_RESET_LABEL, 'default');
	}

	return null;
}

/**
 * HTML-AAM §4.1 (input[type=image]).
 * Name sources: label → alt → title → default ("Submit Query").
 * @see https://www.w3.org/TR/html-aam-1.0/#el-input-image
 */
function handleInputImage(
	el: AccnameElement,
	resolver: AccnameResolver,
	visited: ReadonlySet<string>,
	computeFn: (
		el: AccnameElement,
		resolver: AccnameResolver,
		inLabelledbyTraversal: boolean,
		visited: ReadonlySet<string>,
	) => AccnameResult,
	inLabelledbyTraversal: boolean,
): AccnameResult | null {
	const labelResult = resolveLabelText(el, resolver, visited, computeFn, inLabelledbyTraversal);
	if (labelResult) {
		return labelResult;
	}

	const alt = el.getAttribute('alt');
	if (alt?.trim()) {
		return makeResult(alt, 'alt');
	}

	const title = el.getAttribute('title');
	if (title?.trim()) {
		return makeResult(title, 'title');
	}

	return makeResult(DEFAULT_IMAGE_LABEL, 'default');
}

/**
 * HTML-AAM §4.1 (button).
 * Name sources: label → content → title.
 * @see https://www.w3.org/TR/html-aam-1.0/#el-button
 */
function handleButton(
	el: AccnameElement,
	resolver: AccnameResolver,
	visited: ReadonlySet<string>,
	computeFn: (
		el: AccnameElement,
		resolver: AccnameResolver,
		inLabelledbyTraversal: boolean,
		visited: ReadonlySet<string>,
	) => AccnameResult,
	inLabelledbyTraversal: boolean,
): AccnameResult | null {
	const labelResult = resolveLabelText(el, resolver, visited, computeFn, inLabelledbyTraversal);
	if (labelResult) {
		return labelResult;
	}

	const content = resolveNameFromContent(el, resolver, visited, computeFn, inLabelledbyTraversal);
	if (content.trim()) {
		return makeResult(content, 'content');
	}

	const title = el.getAttribute('title');
	if (title?.trim()) {
		return makeResult(title, 'title');
	}

	return null;
}

/**
 * HTML-AAM §4.1 (fieldset).
 * Name sources: legend (content) → title.
 * @see https://www.w3.org/TR/html-aam-1.0/#el-fieldset
 */
function handleFieldset(
	el: AccnameElement,
	resolver: AccnameResolver,
	visited: ReadonlySet<string>,
	computeFn: (
		el: AccnameElement,
		resolver: AccnameResolver,
		inLabelledbyTraversal: boolean,
		visited: ReadonlySet<string>,
	) => AccnameResult,
	inLabelledbyTraversal: boolean,
): AccnameResult | null {
	const legend = findChildByLocalName(el, 'legend');
	if (legend) {
		const content = resolveNameFromContent(legend, resolver, visited, computeFn, inLabelledbyTraversal);
		if (content.trim()) {
			return makeResult(content, 'legend');
		}
	}

	const title = el.getAttribute('title');
	if (title?.trim()) {
		return makeResult(title, 'title');
	}

	return null;
}

/**
 * HTML-AAM §4.1 (table).
 * Name sources: caption (content) → title.
 * @see https://www.w3.org/TR/html-aam-1.0/#el-table
 */
function handleTable(
	el: AccnameElement,
	resolver: AccnameResolver,
	visited: ReadonlySet<string>,
	computeFn: (
		el: AccnameElement,
		resolver: AccnameResolver,
		inLabelledbyTraversal: boolean,
		visited: ReadonlySet<string>,
	) => AccnameResult,
	inLabelledbyTraversal: boolean,
): AccnameResult | null {
	const caption = findChildByLocalName(el, 'caption');
	if (caption) {
		const content = resolveNameFromContent(caption, resolver, visited, computeFn, inLabelledbyTraversal);
		if (content.trim()) {
			return makeResult(content, 'caption');
		}
	}

	const title = el.getAttribute('title');
	if (title?.trim()) {
		return makeResult(title, 'title');
	}

	return null;
}

/**
 * HTML-AAM §4.1 (img).
 * Name sources: alt → title (only when alt is not specified).
 * When alt="" (empty), returns empty name with null source (decorative image).
 * @see https://www.w3.org/TR/html-aam-1.0/#el-img
 */
function handleImg(el: AccnameElement): AccnameResult | null {
	if (el.hasAttribute('alt')) {
		const alt = el.getAttribute('alt') ?? '';
		return makeResult(alt, alt.trim() ? 'alt' : null);
	}

	const title = el.getAttribute('title');
	if (title?.trim()) {
		return makeResult(title, 'title');
	}

	return null;
}

/**
 * HTML-AAM §4.1 (area).
 * Name sources: alt → title.
 * @see https://www.w3.org/TR/html-aam-1.0/#el-area
 */
function handleArea(el: AccnameElement): AccnameResult | null {
	const alt = el.getAttribute('alt');
	if (alt?.trim()) {
		return makeResult(alt, 'alt');
	}

	const title = el.getAttribute('title');
	if (title?.trim()) {
		return makeResult(title, 'title');
	}

	return null;
}

/**
 * HTML-AAM §4.1 (figure).
 * Name sources: title only. Note: figcaption provides the accessible
 * *description*, not the name, per HTML-AAM.
 * @see https://www.w3.org/TR/html-aam-1.0/#el-figure
 */
function handleFigure(el: AccnameElement): AccnameResult | null {
	const title = el.getAttribute('title');
	if (title?.trim()) {
		return makeResult(title, 'title');
	}

	return null;
}

/**
 * HTML-AAM §4.1 (summary).
 * Name sources: content → title.
 * @see https://www.w3.org/TR/html-aam-1.0/#el-summary
 */
function handleSummary(
	el: AccnameElement,
	resolver: AccnameResolver,
	visited: ReadonlySet<string>,
	computeFn: (
		el: AccnameElement,
		resolver: AccnameResolver,
		inLabelledbyTraversal: boolean,
		visited: ReadonlySet<string>,
	) => AccnameResult,
	inLabelledbyTraversal: boolean,
): AccnameResult | null {
	const content = resolveNameFromContent(el, resolver, visited, computeFn, inLabelledbyTraversal);
	if (content.trim()) {
		return makeResult(content, 'content');
	}

	const title = el.getAttribute('title');
	if (title?.trim()) {
		return makeResult(title, 'title');
	}

	return null;
}

/**
 * HTML-AAM §4.1 (a[href]).
 * Name sources: content → title.
 * @see https://www.w3.org/TR/html-aam-1.0/#el-a
 */
function handleAnchor(
	el: AccnameElement,
	resolver: AccnameResolver,
	visited: ReadonlySet<string>,
	computeFn: (
		el: AccnameElement,
		resolver: AccnameResolver,
		inLabelledbyTraversal: boolean,
		visited: ReadonlySet<string>,
	) => AccnameResult,
	inLabelledbyTraversal: boolean,
): AccnameResult | null {
	const content = resolveNameFromContent(el, resolver, visited, computeFn, inLabelledbyTraversal);
	if (content.trim()) {
		return makeResult(content, 'content');
	}

	const title = el.getAttribute('title');
	if (title?.trim()) {
		return makeResult(title, 'title');
	}

	return null;
}

/**
 * HTML-AAM §4.1 (iframe) — name from title attribute only.
 * @see https://www.w3.org/TR/html-aam-1.0/#el-iframe
 */
function handleTitleOnly(el: AccnameElement): AccnameResult | null {
	const title = el.getAttribute('title');
	if (title?.trim()) {
		return makeResult(title, 'title');
	}

	return null;
}

/**
 * SVG-AAM: SVG elements get name from `<title>` child element.
 * @see https://www.w3.org/TR/svg-aam-1.0/#mapping_additional_nd — SVG-AAM §8.1
 */
function handleSvgElement(el: AccnameElement): AccnameResult | null {
	const titleEl = findChildByLocalName(el, 'title');
	if (titleEl?.textContent?.trim()) {
		return makeResult(titleEl.textContent, 'svg-title');
	}

	return null;
}
